From: Hannes Reinecke <hare@suse.de>
Date: Tue, 10 Sep 2019 14:06:57 +0200
Subject: [PATCH] blk-mq: backport fixes for blk_mq_complete_e_request_sync()
Patch-Mainline: never, backport for SLE15
References: bsc#1145661

Preliminary patches are not ported to SLE15, so we need to
adjust commit
1b8f21b74c3c ("blk-mq: introduce blk_mq_complete_request_sync()")

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 block/blk-mq.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 2fa1753707ba..1ea504b789c0 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -590,8 +590,8 @@ EXPORT_SYMBOL(blk_mq_complete_request);
 
 void blk_mq_complete_request_sync(struct request *rq)
 {
-	WRITE_ONCE(rq->state, MQ_RQ_COMPLETE);
-	rq->q->mq_ops->complete(rq);
+	if (!blk_mark_rq_complete(rq))
+		rq->q->mq_ops->complete(rq);
 }
 EXPORT_SYMBOL_GPL(blk_mq_complete_request_sync);
 
-- 
2.16.4

